cmd/soroban-rpc/internal/xdr2json/conversion.go

```diff
@@ -0,0 +1,144 @@
+//nolint:lll
+package xdr2json
+
+/*
+// See preflight.go for add'l explanations:
+// Note: no blank lines allowed.
+#include <stdlib.h>
+#include "../../lib/preflight.h"
+#include "../../lib/xdrjson.h"
+#cgo windows,amd64 LDFLAGS: -L${SRCDIR}/../../../../target/x86_64-pc-windows-gnu/release-with-panic-unwind/ -lpreflight -lntdll -static -lws2_32 -lbcrypt -luserenv
+#cgo darwin,amd64  LDFLAGS: -L${SRCDIR}/../../../../target/x86_64-apple-darwin/release-with-panic-unwind/ -lpreflight -ldl -lm
+#cgo darwin,arm64  LDFLAGS: -L${SRCDIR}/../../../../target/aarch64-apple-darwin/release-with-panic-unwind/ -lpreflight -ldl -lm
+#cgo linux,amd64   LDFLAGS: -L${SRCDIR}/../../../../target/x86_64-unknown-linux-gnu/release-with-panic-unwind/ -lpreflight -ldl -lm
+#cgo linux,arm64   LDFLAGS: -L${SRCDIR}/../../../../target/aarch64-unknown-linux-gnu/release-with-panic-unwind/ -lpreflight -ldl -lm
+*/
+import "C"
+
+import (
+	"encoding"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+	"unsafe"
+
+	"github.com/pkg/errors"
+
+	"github.com/stellar/go/xdr"
+
+	"github.com/stellar/soroban-rpc/cmd/soroban-rpc/internal/db"
+)
+
+const (
+	FormatBase64 = "base64"
+	FormatJSON   = "json"
+)
+
+var errInvalidFormat = fmt.Errorf(
+	"expected %s for optional 'xdrFormat'",
+	strings.Join([]string{FormatBase64, FormatJSON}, ", "))
+
+// ConvertBytes takes an XDR object (`xdr`) and its serialized bytes (`field`)
+// and returns the raw JSON-formatted serialization of that object.
+// It can be unmarshalled to a proper JSON structure, but the raw bytes are
+// returned to avoid unnecessary round-trips. If there is an
+// error, it returns an empty JSON object.
+//
+// The `xdr` object does not need to actually be initialized/valid:
+// we only use it to determine the name of the structure. We could just
+// accept a string, but that would make mistakes likelier than passing the
+// structure itself (by reference).
+func ConvertBytes(xdr interface{}, field []byte) ([]byte, error) {
+	xdrTypeName := reflect.TypeOf(xdr).Name()
+	return convertAnyBytes(xdrTypeName, field)
+}
+
+// ConvertInterface takes a valid XDR object (`xdr`) and returns
+// the raw JSON-formatted serialization of that object. If there is an
+// error, it returns an empty JSON object.
+//
+// Unlike `ConvertBytes`, the value here needs to be valid and
+// serializable.
+func ConvertInterface(xdr interface{}) (json.RawMessage, error) {
```

Comment:
If you expect a `BinaryMarshaler` why not enforce so directly?

Although, wouldn't it make more sense to enforce `xdr.EncodeTo`? 

Also, consider using an `xdr.Encoder` directly where possible to make encoding more efficient (less garbage(

Reply:
If the problem is you need the type name and you still need reflection, you can still do it internally, without exposing an `interface{}`

Reply:
But note that reflection is slow and has a high performance impact.

Reply:
If the typename is strictly neccessary we can modify `xdrgen` to provide it thourgh a method.

Reply:
Nice, not sure why I didn't realize that I could just ask for the interface I wanted. I think `BinaryMarshaler` is fine since the only thing I need is `MarshalBinary`, so it feels prescriptive to require a higher-level interface.

EncodeTo is a good idea, but it will make caller code more verbose. I'll explore this to see if it's worthwhile! It may not be since each structure is separate in memory.

> But note that reflection is slow and has a high performance impact.
> If the typename is strictly neccessary we can modify xdrgen to provide it thourgh a method.

It is necessary, yeah. Totally agreed that it's better to have a `Name()` method on each XDR structure that gets generated; using reflection was the only way to get the name otherwise. Since we'd need to modify xdrgen, we should do it after this PR and patch it later; it'll be a small fix here.

Reply:
I added a benchmark in 1ef0940, here are the results:

```
$ go test -bench=. -run='BenchmarkJSONTransactions' ./cmd/soroban-rpc/internal/methods/...
goos: linux
goarch: amd64
pkg: github.com/stellar/soroban-rpc/cmd/soroban-rpc/internal/methods
cpu: 13th Gen Intel(R) Core(TM) i7-1360P
BenchmarkJSONTransactions/JSON_format-16         	  199840	      6885 ns/op
BenchmarkJSONTransactions/XDR_format-16          	  709465	      2500 ns/op
```

You can see that JSON mode is ~4x slower than XDR mode, but this is not due to reflection. In fact, reflection is not even in the top 100 for the CPU trace. Profiling shows that the **additional time is spent almost exclusively in the cgocall** across the Go -> C (Rust) boundary:

```
(pprof) top5  
Showing nodes accounting for 1930ms, 42.05% of 4590ms total
Dropped 138 nodes (cum <= 22.95ms)
Showing top 5 nodes out of 156
      flat  flat%   sum%        cum   cum%
    1190ms 25.93% 25.93%     1510ms 32.90%  runtime.cgocall
     230ms  5.01% 30.94%      230ms  5.01%  runtime.duffcopy
     220ms  4.79% 35.73%      660ms 14.38%  runtime.mallocgc
     160ms  3.49% 39.22%      160ms  3.49%  encoding/base64.(*Encoding).Encode
     130ms  2.83% 42.05%      320ms  6.97%  runtime.scanobject
```

<hr>

Side note for @leighmcculloch: I wanted to see how much your suggestion about adding the `json.RawMessage` support helped. It led to a ~2.5x speedup since there's much less time doing marshaling! So you rock :guitar: Details:

<details>

```
$ go test -bench=. -run='BenchmarkJSONTransactions' ./cmd/soroban-rpc/internal/methods/...
goos: linux
goarch: amd64
pkg: github.com/stellar/soroban-rpc/cmd/soroban-rpc/internal/methods
cpu: 13th Gen Intel(R) Core(TM) i7-1360P
BenchmarkJSONTransactions/JSON_format-16         	   69666	     16455 ns/op
BenchmarkJSONTransactions/XDR_format-16          	  515062	      2454 ns/op
```

</details>

Reply:
Following up from Slack. I ran a Python benchmark that queried for random transactions 5000 times. Here are the latency results:

```bash
$ python3 benchmark.py 
5000 queries with xdrFormat=base64 format took 7.89s
5000 queries with xdrFormat=json format took 11.33s
$ python3 benchmark.py 
5000 queries with xdrFormat=base64 format took 7.89s
5000 queries with xdrFormat=json format took 11.66s
$ python3 benchmark.py 
5000 queries with xdrFormat=base64 format took 8.19s
5000 queries with xdrFormat=json format took 12.00s
```

As you can see, JSON mode definitely does take noticeably longer (about ~40% overhead), but it's much less overhead when amortized over the entire request.